spec.md

```diff
@@ -1713,76 +1725,97 @@ In the example above, the type of `mymem`{.firrtl} is:
 {flip r1: {addr: UInt<8>,
            en: UInt<1>,
            clk: Clock,
-           flip data: {real: SInt<16>, imag: SInt<16>}},
+           flip rdata: {real: SInt<16>, imag: SInt<16>}},
  flip r2: {addr: UInt<8>,
            en: UInt<1>,
            clk: Clock,
-           flip data: {real: SInt<16>, imag: SInt<16>}},
+           flip rdata: {real: SInt<16>, imag: SInt<16>}},
  flip w: {addr: UInt<8>,
           en: UInt<1>,
           clk: Clock,
-          data: {real: SInt<16>, imag: SInt<16>},
-          mask: {real: UInt<1>, imag: UInt<1>}}}
+          wdata: {real: SInt<16>, imag: SInt<16>},
+          mask: {real: UInt<1>, imag: UInt<1>}},
+ custom: {a:UInt<4>, flip b:UInt<2>}}
 ```
 
 The following sections describe how a memory's field types are calculated and
 the behavior of each type of memory port.
 
-### Read Ports
+### Ports
 
-If a memory is declared with element type `T`{.firrtl}, has a size less than or
-equal to $2^N$, then its read ports have type:
+Ports can have one of the following read capabilities:
 
-``` firrtl
-{addr: UInt<N>, en: UInt<1>, clk: Clock, flip data: T}
-```
+- `no`{.firrtl}: This port cannot read from the memory
+- `yes`{.firrtl}: This port can read from the memory
 
-If the `en`{.firrtl} field is high, then the element value associated with the
-address in the `addr`{.firrtl} field can be retrieved by reading from the
-`data`{.firrtl} field after the appropriate read latency. If the `en`{.firrtl}
-field is low, then the value in the `data`{.firrtl} field, after the appropriate
-read latency, is undefined. The port is driven by the clock signal in the
-`clk`{.firrtl} field.
+Ports can have one of the following write capabilities:
 
-### Write Ports
+- `no`{.firrtl}: This port cannot write into the memory
+- `no-mask`{.firrtl}: This port can only do full writes into the memory
+- `with-mask`{.firrtl}: This port can do partial writes into the memory
 
 If a memory is declared with element type `T`{.firrtl}, has a size less than or
-equal to $2^N$, then its write ports have type:
+equal to $2^N$, then a port with maximum capability (
+`read => yes, write => with-mask`{.firrtl}) has type:
 
 ``` firrtl
-{addr: UInt<N>, en: UInt<1>, clk: Clock, data: T, mask: M}
+{
+  clk: Clock,
+  en: UInt<1>,
+  addr: UInt<N>,
+  wmode: UInt<1>,
+
+  flip rdata: T,
+  wdata: T,
+  wmask: M
+}
 ```
 
 where `M`{.firrtl} is the mask type calculated from the element type
```

Comment:
Does the mask actually help anywhere?  It seems it is a feature which really only makes sense assuming a certain compilation strategy (lowering aggregate memories to multiple memories, in which the mask becomes the enable flag).

Reply:
IMO, this is a abstraction over different possible hardware memory implementations (some SRAM allows bit or byte masks). Aggregated memory lowering might be part of  memory synthesis, which one might want to delay until backend flow.

Also, this might be good for behavioral simulations performance?